File: examples/java/03-functions/Main.java
import java.util.*;
import java.util.function.*;
import java.util.stream.*;

/**
 * Functions in Java
 * =================
 * Java added lambdas and functional interfaces in Java 8 (2014).
 * Before that, anonymous inner classes were the only way to pass
 * behaviour. Java functions are NOT first-class in the traditional
 * sense — lambdas are syntactic sugar for functional interface
 * implementations.
 */
public class Main {
    public static void main(String[] args) {

        // ── Basic methods ────────────────────────────

        System.out.println(greet("Atlas"));

        // ── Lambda expressions (Java 8+) ─────────────

        System.out.println("\n--- Lambda expressions ---");

        // Functional interface: an interface with one abstract method
        // Java provides many: Function, Predicate, Consumer, Supplier, etc.

        Function<Integer, Integer> square = x -> x * x;
        Function<Integer, Integer> doubleIt = x -> x * 2;

        System.out.println("square(5) = " + square.apply(5));
        System.out.println("doubleIt(5) = " + doubleIt.apply(5));

        // Predicate: T -> boolean
        Predicate<Integer> isEven = x -> x % 2 == 0;
        System.out.println("isEven(4) = " + isEven.test(4));
        System.out.println("isEven(5) = " + isEven.test(5));

        // BiFunction: (T, U) -> R
        BiFunction<Integer, Integer, Integer> add = (a, b) -> a + b;
        System.out.println("add(3, 4) = " + add.apply(3, 4));

        // ── Method references ────────────────────────

        System.out.println("\n--- Method references ---");

        // Instead of lambda: x -> x.toUpperCase()
        Function<String, String> upper = String::toUpperCase;
        System.out.println("upper(\"hello\") = " + upper.apply("hello"));

        // Static method reference
        Function<String, Integer> parse = Integer::parseInt;
        System.out.println("parse(\"42\") = " + parse.apply("42"));

        // ── Streams (higher-order functions) ─────────

        System.out.println("\n--- Streams ---");

        List<Integer> numbers = List.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);

        // map
        List<Integer> squares = numbers.stream()
                .map(x -> x * x)
                .collect(Collectors.toList());
        System.out.println("squares = " + squares);

        // filter
        List<Integer> evens = numbers.stream()
                .filter(x -> x % 2 == 0)
                .collect(Collectors.toList());
        System.out.println("evens = " + evens);

        // reduce (fold)
        int total = numbers.stream()
                .reduce(0, Integer::sum);
        System.out.println("sum = " + total);

        int product = numbers.stream()
                .reduce(1, (a, b) -> a * b);
        System.out.println("product = " + product);

        // Chaining
        int sumOfSquaresOfEvens = numbers.stream()
                .filter(x -> x % 2 == 0)
                .map(x -> x * x)
                .reduce(0, Integer::sum);
        System.out.println("sum of squares of evens = " + sumOfSquaresOfEvens);

        // Collectors
        String joined = List.of("hello", "functional", "java").stream()
                .map(String::toUpperCase)
                .collect(Collectors.joining(", "));
        System.out.println("joined = " + joined);

        // any/all
        boolean hasNegative = numbers.stream().anyMatch(x -> x < 0);
        boolean allPositive = numbers.stream().allMatch(x -> x > 0);
        System.out.println("any negative? " + hasNegative);
        System.out.println("all positive? " + allPositive);

        // ── Function composition ─────────────────────

        System.out.println("\n--- Function composition ---");

        Function<Integer, Integer> addOne = x -> x + 1;
        Function<Integer, Integer> triple = x -> x * 3;

        // andThen: first this, then that
        Function<Integer, Integer> addOneThenTriple = addOne.andThen(triple);
        System.out.println("addOne.andThen(triple)(5) = " + addOneThenTriple.apply(5));

        // compose: first that, then this
        Function<Integer, Integer> tripleComposeAddOne = triple.compose(addOne);
        System.out.println("triple.compose(addOne)(5) = " + tripleComposeAddOne.apply(5));

        // ── Closures (capturing variables) ───────────

        System.out.println("\n--- Closures ---");

        int factor = 3;
        // Lambdas can capture effectively final variables
        Function<Integer, Integer> timesThree = x -> x * factor;
        System.out.println("timesThree(5) = " + timesThree.apply(5));

        // factor = 4;  // ERROR: variable used in lambda must be effectively final

        // Returning functions (factory pattern)
        Function<Integer, Integer> timesTwo = makeMultiplier(2);
        Function<Integer, Integer> timesFive = makeMultiplier(5);
        System.out.println("timesTwo(7) = " + timesTwo.apply(7));
        System.out.println("timesFive(7) = " + timesFive.apply(7));

        // ── Optional (functional null handling) ──────

        System.out.println("\n--- Optional ---");

        Optional<String> present = Optional.of("hello");
        Optional<String> absent = Optional.empty();

        System.out.println("present.map(upper) = " + present.map(String::toUpperCase));
        System.out.println("absent.map(upper) = " + absent.map(String::toUpperCase));
        System.out.println("absent.orElse(\"default\") = " + absent.orElse("default"));

        // Chaining Optionals
        String result = Optional.of("42")
                .map(Integer::parseInt)
                .filter(n -> n > 0)
                .map(n -> "Positive: " + n)
                .orElse("Not positive");
        System.out.println("Optional chain = " + result);

        // ── Custom higher-order function ─────────────

        System.out.println("\n--- Custom HOF ---");

        System.out.println("applyTwice(x+1, 5) = " + applyTwice(x -> x + 1, 5));
        System.out.println("applyTwice(x*2, 3) = " + applyTwice(x -> x * 2, 3));

        // ── Sorting with comparators ─────────────────

        System.out.println("\n--- Sorting ---");

        List<String> words = new ArrayList<>(List.of("banana", "apple", "cherry", "date"));

        words.sort(Comparator.comparingInt(String::length));
        System.out.println("by length = " + words);

        words.sort(Comparator.naturalOrder());
        System.out.println("alphabetical = " + words);

        words.sort(Comparator.comparingInt(String::length).reversed());
        System.out.println("by length desc = " + words);

        // ── Summary ──────────────────────────────────

        System.out.println("\n--- Summary ---");
        System.out.println("Java functions:");
        System.out.println("  - Lambdas since Java 8 (2014)");
        System.out.println("  - Functional interfaces: Function, Predicate, Consumer");
        System.out.println("  - Method references: String::toUpperCase");
        System.out.println("  - Streams: map, filter, reduce, collect");
        System.out.println("  - andThen/compose for function composition");
        System.out.println("  - Closures capture effectively final variables only");
        System.out.println("  - Optional for functional null handling");
    }

    static String greet(String name) {
        return "Hello, " + name + "!";
    }

    static Function<Integer, Integer> makeMultiplier(int factor) {
        return x -> x * factor;
    }

    static <T> T applyTwice(Function<T, T> f, T x) {
        return f.apply(f.apply(x));
    }
}

File: examples/go/03-functions/main.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

// ── Basic function ───────────────────────────────────

func greet(name string) string {
	return "Hello, " + name + "!"
}

// ── Multiple return values ───────────────────────────

func divide(a, b float64) (float64, error) {
	if b == 0 {
		return 0, fmt.Errorf("division by zero")
	}
	return a / b, nil
}

// Named return values
func minMax(nums []int) (min, max int) {
	min, max = nums[0], nums[0]
	for _, n := range nums[1:] {
		if n < min {
			min = n
		}
		if n > max {
			max = n
		}
	}
	return // naked return — returns named values
}

// ── Variadic functions ───────────────────────────────

func sum(nums ...int) int {
	total := 0
	for _, n := range nums {
		total += n
	}
	return total
}

// ── Higher-order functions ───────────────────────────

func apply(f func(int) int, values []int) []int {
	result := make([]int, len(values))
	for i, v := range values {
		result[i] = f(v)
	}
	return result
}

func filter(pred func(int) bool, values []int) []int {
	var result []int
	for _, v := range values {
		if pred(v) {
			result = append(result, v)
		}
	}
	return result
}

func reduce(f func(int, int) int, values []int, initial int) int {
	acc := initial
	for _, v := range values {
		acc = f(acc, v)
	}
	return acc
}

// ── Closures ─────────────────────────────────────────

func makeMultiplier(factor int) func(int) int {
	return func(x int) int {
		return x * factor // factor captured from enclosing scope
	}
}

func makeCounter() func() int {
	count := 0
	return func() int {
		count++ // mutates captured variable
		return count
	}
}

// ── Function composition ─────────────────────────────

func compose(f, g func(string) string) func(string) string {
	return func(s string) string {
		return f(g(s))
	}
}

// ── Methods (functions with receivers) ───────────────

type Point struct {
	X, Y float64
}

// Value receiver — doesn't modify original
func (p Point) Scale(factor float64) Point {
	return Point{p.X * factor, p.Y * factor}
}

// String method — satisfies fmt.Stringer interface
func (p Point) String() string {
	return fmt.Sprintf("(%g, %g)", p.X, p.Y)
}

// ── Defer ────────────────────────────────────────────

func demonstrateDefer() {
	fmt.Println("  start")
	defer fmt.Println("  deferred 1 (runs last)")
	defer fmt.Println("  deferred 2 (runs second-to-last)")
	fmt.Println("  end")
	// Deferred calls execute in LIFO order when function returns
}

// ── Main ─────────────────────────────────────────────

func main() {
	// Basic function
	fmt.Println(greet("Atlas"))

	// Multiple return values
	fmt.Println("\n--- Multiple return values ---")
	result, err := divide(10, 3)
	if err != nil {
		fmt.Println("Error:", err)
	} else {
		fmt.Printf("10 / 3 = %.4f\n", result)
	}

	_, err = divide(10, 0)
	if err != nil {
		fmt.Println("Error:", err)
	}

	min, max := minMax([]int{3, 1, 4, 1, 5, 9, 2, 6})
	fmt.Printf("min = %d, max = %d\n", min, max)

	// Variadic functions
	fmt.Println("\n--- Variadic ---")
	fmt.Println("sum(1,2,3) =", sum(1, 2, 3))
	fmt.Println("sum(1..10) =", sum(1, 2, 3, 4, 5, 6, 7, 8, 9, 10))
	nums := []int{10, 20, 30}
	fmt.Println("sum(slice...) =", sum(nums...)) // spread slice

	// First-class functions
	fmt.Println("\n--- First-class functions ---")
	add := func(a, b int) int { return a + b }
	fmt.Println("add(3, 4) =", add(3, 4))

	// Higher-order functions
	fmt.Println("\n--- Higher-order functions ---")
	numbers := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}

	squares := apply(func(x int) int { return x * x }, numbers)
	fmt.Println("squares =", squares)

	evens := filter(func(x int) bool { return x%2 == 0 }, numbers)
	fmt.Println("evens =", evens)

	total := reduce(func(a, b int) int { return a + b }, numbers, 0)
	fmt.Println("sum =", total)

	product := reduce(func(a, b int) int { return a * b }, numbers, 1)
	fmt.Println("product =", product)

	// Sorting with custom comparator
	words := []string{"banana", "apple", "cherry", "date"}
	sort.Slice(words, func(i, j int) bool {
		return len(words[i]) < len(words[j])
	})
	fmt.Println("sorted by length =", words)

	// Closures
	fmt.Println("\n--- Closures ---")
	double := makeMultiplier(2)
	triple := makeMultiplier(3)
	fmt.Println("double(5) =", double(5))
	fmt.Println("triple(5) =", triple(5))

	counter := makeCounter()
	fmt.Println("counter() =", counter())
	fmt.Println("counter() =", counter())
	fmt.Println("counter() =", counter())

	// Function composition
	fmt.Println("\n--- Composition ---")
	shout := compose(strings.ToUpper, strings.TrimSpace)
	fmt.Printf("shout(\"  hello  \") = %q\n", shout("  hello  "))

	// Methods
	fmt.Println("\n--- Methods ---")
	p := Point{3, 4}
	fmt.Println("p =", p)
	fmt.Println("p.Scale(2) =", p.Scale(2))
	fmt.Println("p (unchanged) =", p) // value receiver doesn't modify

	// Defer
	fmt.Println("\n--- Defer ---")
	demonstrateDefer()

	// Immediately invoked function
	fmt.Println("\n--- IIFE ---")
	result2 := func(x int) int { return x * x }(7)
	fmt.Println("(func(x) x*x)(7) =", result2)

	// Summary
	fmt.Println("\n--- Summary ---")
	fmt.Println("Go functions:")
	fmt.Println("  - First-class: assignable, passable")
	fmt.Println("  - Multiple return values (idiomatic error handling)")
	fmt.Println("  - Variadic: func(args ...int)")
	fmt.Println("  - Closures: capture enclosing scope")
	fmt.Println("  - Methods: functions with receivers")
	fmt.Println("  - Defer: cleanup runs when function returns")
	fmt.Println("  - No generics for HOF (until 1.18+)")
}
